Exercise 1 branch

muzaluisa opened the pull request:
Hey, could you please review my first assignment? :)
It is plug & play, but you can add your own test as well easily.

muzaluisa committed:
Two solutions for the first week exercise!

muzaluisa committed:
Assignment 1 description

muzaluisa committed:
Update README.md

muzaluisa commented:
Hey, 
Could you please review my first exercise?)

luxiya01 commented on codeU1.py:
```diff
@@ -0,0 +1,103 @@
+        return 1;
+        
+    
+is_permutation("permutation","")
```
Perhaps you can split out the tests into a test file called "test.py" or something? I feel the way the tests are presented in the code makes the code a bit messy right now.

luxiya01 commented on codeU1.py:
```diff
@@ -0,0 +1,103 @@
+        cur = self.head;
+        for i in range(0,self.N-k-1):
+            cur = cur.next
+        return cur.val;    
```
This function seems to have a misleading name. It seems to return the value stored in the k:th node in the linked list, but the name "k_last" doesn't really reflect this. Perhaps you can change it to "get_kth_node", "get_kth_element" or something?

muzaluisa replied:
Great, I have renamed the function and made separate test functions :)

luxiya01 reviewed (commented):
Just some comments about the styling...

muzaluisa committed:
Changes according to @luxiya01

Thank you for the review
Renamed function to get_kth_last_element + made tests in the separate function :)

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+"""
+
+# Task 1: Given 2 strings, decide if one of them is a permutation of the other
+# Return:
```
Documenting the behavior is nice - makes it easier to use the library and serves as a lightweight contract between the library and users.

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+    str1 = str1.strip().lower()
+    str2 = str2.strip().lower()
+    if  not str1 and not str2:
+        print "Two empty strings are permutations"
```
Also maybe take a look at python logging library:

https://docs.python.org/3.6/howto/logging.html#logging-basic-tutorial

It is not at all important here, but in general one would use logging to output diagnostic information rather than direct printing.

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+def is_permutation(str1,str2):
+    yes = "'" + str1 + "'" + " is a permutation of " + "'" + str2 + "'";
+    no = "'" + str1 + "'" + " is not a permutation of " + "'" + str2 + "'";
+    str1 = str1.strip().lower()
```
This depends on the use case of course, but sometimes leading whitespace (and case) can be important. In production code, it would be important to document that this string manipulation is taking place.

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+        return 1
+    ord1 = [ord(a) for a in str1]
+    ord2 = [ord(a) for a in str2]
+    if np.prod(ord1)!=np.prod(ord2):
```
Would you expect the product and sum to never misidentify non-permutations as permutations?

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+        print yes
+        return 1;
+        
+def test_permutation():    
```
Adding tests is good, consider using the unittest library for these:

https://docs.python.org/3.6/library/unittest.html?highlight=unittest#basic-example

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+import numpy as np
+   
+def is_permutation(str1,str2):
+    yes = "'" + str1 + "'" + " is a permutation of " + "'" + str2 + "'";
```
Nit-picky comment: No semicolons required in Python code.

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+        new_node = Node(value)
+        if not self.head:
+            self.head = new_node;
+            self.N = self.N+1;
```
You can also do this in Python:

self.N += 1

(Though there is no self.N++ in Python)

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+            self.head = new_node;
+            self.N = self.N+1;
+            return;
+        cur = self.head
```
This is a completely valid way to do an insert, but just as a follow-up question: Is there a way to implement a constant-time insert() for linked lists (taking different tradeoffs)?

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+
+    def get_kth_last_element(self,k):
+        if k > self.N-1:
+            print 'There are ' + str(self.N) + ' elements, k should be less than N';
```
In Python, exceptions are generally employed when basic constraints are violated:

```
>>> a = [1, 2, 3]
>>> a[4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
````

siimphh commented on codeU1.py:
```diff
@@ -0,0 +1,118 @@
+            print 'The list is empty';
+            return None;
+        cur = self.head;
+        for i in range(0,self.N-k-1):
```
AFAICT this is indeed the most efficient way solving the problem on linked lists.

siimphh reviewed (commented):
Some post-merge comments for your considerations. Note that it is OK (even normal) to not agree with all the comments :)
